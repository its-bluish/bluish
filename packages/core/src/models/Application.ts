import { Middleware } from './Middleware.js'

export abstract class Application extends Middleware {}
