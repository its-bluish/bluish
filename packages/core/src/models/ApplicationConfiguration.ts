/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/prefer-ts-expect-error */
import { Host } from './Host'
import { HookCollection } from './HookCollection'
import { PluginCollection } from './PluginCollection'
import { CoreEmitter } from '../helpers/CoreEmitter'

global.bluishApplicationConfiguration = null

export class ApplicationConfiguration {
  public static has() {
    return !!globalThis.bluishApplicationConfiguration
  }

  public static get() {
    return globalThis.bluishApplicationConfiguration ?? null
  }

  public static isSame(target: Function | Object) {
    const constructor = typeof target === 'function' ? target : target.constructor

    return constructor === globalThis.bluishApplicationConfiguration?.target
  }

  public static set(target: Function | Object | null): ApplicationConfiguration {
    if (target === null) return new this(null as unknown as Function)

    if (typeof target !== 'function') return this.set(target.constructor)

    if (globalThis.bluishApplicationConfiguration?.target !== target) {
      const applicationConfiguration: ApplicationConfiguration = new this(target)

      // @ts-ignore
      globalThis.bluishApplicationConfiguration = applicationConfiguration

      CoreEmitter.emit('application-configuration', applicationConfiguration)
    }

    // @ts-ignore
    return globalThis.bluishApplicationConfiguration
  }

  public hooks = new HookCollection()

  public plugins = new PluginCollection()

  public host = new Host()

  protected constructor(public target: Function) {}
}

declare global {
  // @ts-ignore
  // eslint-disable-next-line no-inner-declarations, vars-on-top, no-var
  var bluishApplicationConfiguration: ApplicationConfiguration | null
}
