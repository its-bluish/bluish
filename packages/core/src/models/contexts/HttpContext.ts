/* eslint-disable @typescript-eslint/prefer-ts-expect-error */
/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable object-shorthand */
import { Context, AzureFunctionContext } from './Context'
import { HttpRequest } from '@azure/functions'
import { PromiseToo } from '../../typings/PromiseToo'

export type { HttpRequest as AzureHttpRequest } from '@azure/functions'

export type HttpMethod = Lowercase<
  'GET' | 'POST' | 'DELETE' | 'HEAD' | 'PATCH' | 'PUT' | 'OPTIONS' | 'TRACE' | 'CONNECT'
>

export interface HttpTemplate {
  query: Record<string, unknown>
  params: Record<string, unknown>
  body: unknown
}

interface Response {
  status?: number
  headers: Record<string, string>
}
export class HttpContext<T extends Partial<HttpTemplate> = {}> extends Context {
  public method: HttpMethod
  public url: string
  public headers: Record<string, string>
  public query: HttpTemplate['query'] & T['query']
  public params: HttpTemplate['params'] & T['params']
  public body: T['body']
  public rawBody: unknown
  private _response: Response = {
    headers: {},
  }

  constructor(context: AzureFunctionContext, public azureFunctionRequest: HttpRequest) {
    super(context)

    this.method = azureFunctionRequest.method?.toLowerCase() as HttpMethod
    this.url = azureFunctionRequest.url
    this.headers = Object.fromEntries(
      Object.entries(azureFunctionRequest.headers).flatMap(([key, value]) => [
        [key, value],
        [key.toLowerCase(), value],
      ]),
    )
    this.query = azureFunctionRequest.query
    this.params = azureFunctionRequest.params
    this.body = azureFunctionRequest.body as unknown
    this.rawBody = azureFunctionRequest.rawBody
  }

  protected buildResponsePayload(payload: unknown) {
    const { status, headers } = this._response

    if (!payload) return { status, headers, body: payload }

    if (typeof payload !== 'object') return { status, headers, body: payload }

    if ('status' in payload && 'body' in payload) {
      return {
        // @ts-ignore
        status: payload.status,
        // @ts-ignore
        body: payload.body,
        headers: {
          ...headers,
          // @ts-ignore
          ...('headers' in payload && typeof payload.headers === 'object' && payload.headers),
        },
      }
    }

    return { status, headers, body: payload }
  }

  public success(payload: unknown): PromiseToo<unknown> {
    this.azureFunctionContext.res = this.buildResponsePayload(payload)
    if (!this.azureFunctionContext.res.status) this.azureFunctionContext.res.status = 200
    return void 0
  }

  public unhandledError(error: unknown): PromiseToo<unknown> {
    throw error
  }

  public handledError(payload: unknown): PromiseToo<unknown> {
    this.azureFunctionContext.res = this.buildResponsePayload(payload)
    if (!this.azureFunctionContext.res.status) this.azureFunctionContext.res.status = 500
    return void 0
  }

  public status(): number
  public status(status: number): this
  public status(status?: number) {
    if (status === void 0) return this._response.status

    this._response.status = status

    return this
  }

  public setHeader(name: string, value: string): this
  public setHeader(name: string, value: string) {
    this._response.headers[name] = value

    return this
  }
}

export interface HttpContext<T extends Partial<HttpTemplate> = {}>
  extends Context,
    Bluish.HttpContext<T> {}

declare global {
  export namespace Bluish {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    export interface HttpContext<T extends Partial<HttpTemplate>> {}
  }
}
