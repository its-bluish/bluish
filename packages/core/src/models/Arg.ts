import { Context } from './contexts/Context'
import { TriggerConfiguration } from './TriggerConfiguration'

export type ArgFactory<C extends Context> = (context: C) => unknown

export class Arg<C extends Context> {
  public trigger!: TriggerConfiguration

  constructor(public factory: ArgFactory<C>) {}
}
