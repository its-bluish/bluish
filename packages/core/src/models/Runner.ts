import { Trigger } from './metadata/Trigger';
import { ApplicationMetadata } from './metadata/Application';
import { Hook } from './metadata/Hook';
import { AzureFunctionContext, Context } from "./contexts/Context";
import { Metadata } from './metadata';
import { wait } from '../tools/wait';

type Class<R = any, A extends any[] = any[]> = new (...args: A) => R

export class Runner {
  public instance!: Record<string, Function>
  public application: Record<string, Function>
  public applicationMetadata: ApplicationMetadata
  public trigger: Trigger
  public metadata: Metadata

  constructor(
    public targetConstructor: Class,
    public property: string,
    public applicationConstructor: Class | null = null
  ) {
    this.application = applicationConstructor ? new applicationConstructor() : {}

    this.applicationMetadata = ApplicationMetadata.set(applicationConstructor)

    this.metadata = Metadata.loadOrFail(targetConstructor)

    this.trigger = this.metadata.triggers.findOneByPropertyOrFail(property)

    this.instance = new targetConstructor()
  }

  protected get plugins() {
    return [
      ...this.applicationMetadata.plugins,
      ...this.metadata.plugins,
      ...this.trigger.plugins,
    ]
  }

  public async run(context: Context) {
    return await this.instance[this.property].apply(
      this.instance,
      await this.trigger.args.toArguments(context)
    )
  }

  private async _hook(target: Record<string, Function>, hook: Hook, ...args: unknown[]) {
    if (typeof hook.call === 'function') return await hook.call(...args)

    return await target[hook.call].call(target, ...args)
  }

  private async _hooks(event: string, ...args: unknown[]) {
    for (const hook of this.applicationMetadata.hooks.findByEvent(event))

      await this._hook(this.application, hook, ...args)

    for (const hook of this.metadata.hooks.findByEvent(event))

      await this._hook(this.instance, hook, ...args)

    for (const hook of this.trigger.hooks.findByEvent(event))

      await this._hook({}, hook, ...args)
  }

  private async _initialize(context: Context) {
    await context.initialize()

    for (const plugin of this.plugins) await plugin.onInitialize(context)

    await this._hooks('initialize', context)
  }

  private async _destroy(context: Context) {
    for (const plugin of this.plugins) await plugin.onDestroy(context)

    await this._hooks('destroy', context)

    await context.destroy()
  }

  private async _error(error: unknown, context: Context) {
    for (const plugin of this.plugins) await plugin.onError(error, context)

    await this._hooks('error', error, context)
  }

  private async _handleError(error: unknown, context: Context) {
    for (const hook of this.trigger.hooks.findByEvent('error-handler')) {
      const data = await this._hook({}, hook, error, context)

      if (data !== undefined) return data
    }

    for (const hook of this.metadata.hooks.findByEvent('error-handler')) {
      const data = await this._hook(this.instance, hook, error, context)

      if (data !== undefined) return data
    }

    for (const hook of this.applicationMetadata.hooks.findByEvent('error-handler')) {
      const data = await this._hook(this.application, hook, error, context)

      if (data !== undefined) return data
    }
  }

  private async _transform(data: unknown, context: Context) {
    data = await this.plugins
      .reduce(async (data, plugin) => await plugin.transform(await data, context) ?? data, Promise.resolve(data))

    data = await this.metadata
      .hooks
      .findByEvent('transform')
      .reduce(async (data, hook) => await this._hook(this.instance, hook, await data, context) ?? data, Promise.resolve(data))

    data = await this.trigger
      .hooks
      .findByEvent('transform')
      .reduce(async (data, hook) => await this._hook({}, hook, await data, context) ?? data, Promise.resolve(data))

    return data
  }

  private async _run(azureFunctionContext: AzureFunctionContext, ...args: unknown[]) {
    await wait.cleaning()

    const context = Object.assign(
      new this.trigger.Context(azureFunctionContext, ...args),
      { runner: this }
    )

    try {
      await this._initialize(context)

      return await context.success(
        await this._transform(
          await this.run(context),
          context
        )
      )
    } catch (error) {
      await this._error(error, context)

      const maybeHandledError = await this._handleError(error, context)

      if (maybeHandledError !== undefined)

        return await context.handledError(maybeHandledError)

      return await context.unhandledError(error)
    } finally {
      await this._destroy(context)
    }
  }

  public toAzureFunction() {
    return async (azureFunctionContext: AzureFunctionContext, ...args: unknown[]) => {
      return await this._run(azureFunctionContext, ...args)
    }
  }
}
