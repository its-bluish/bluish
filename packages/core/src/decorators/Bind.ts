import { Context } from '../models/contexts/Context'
import { Arg, ArgFactory } from '../models/Arg'
import { waitForTriggerConfiguration } from '../tools/waitForTriggerConfiguration'

export function Bind<C extends Context>(factory: ArgFactory<C> = (context) => context) {
  return (target: Object, property: string, index: number) => {
    waitForTriggerConfiguration(target, property, (triggerConfiguration) => {
      triggerConfiguration.args.set(index, new Arg<C>(factory) as Arg<Context>)
    })
  }
}
