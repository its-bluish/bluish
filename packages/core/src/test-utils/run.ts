import { AzureFunctionContext } from '../models/contexts/Context'
import { Runner } from '../models/Runner'

export function run(runner: Runner, context: Partial<AzureFunctionContext> = {}, arg: any = {}) {
  return runner.toAzureFunction()((context ?? {}) as AzureFunctionContext, arg)
}
