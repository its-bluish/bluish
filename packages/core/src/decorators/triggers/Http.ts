import { Trigger } from './Trigger'
import { Binding } from '../../models/Binding'
import { HttpContext, HttpMethod } from '../../models/contexts/HttpContext'
import { Bind } from '../Bind'
import { TriggerDecorator, TriggerParamDecorator } from '../../typings/decorators'
import { OnSuccess } from '../OnSuccess'

const HTTP_ROUTE_NORMALIZER_REGEXP = /^\/+|\/+$/

export interface HttpOptions<C extends HttpContext> {
  name?: string
  Context?: new (...args: any[]) => C
  type?: 'binary' | 'stream' | 'string'
}

export function Http<C extends HttpContext>(
  route: string,
  methods: HttpMethod | HttpMethod[],
  { Context = HttpContext as new (...args: any[]) => C, type }: HttpOptions<C> = {},
): TriggerDecorator {
  const bindings = [
    new Binding('httpTrigger', 'req', 'in', {
      route: route.replace(HTTP_ROUTE_NORMALIZER_REGEXP, ''),
      methods: Array.isArray(methods) ? methods : [methods],
      dataType: type,
    }),
    new Binding('http', 'res', 'out'),
  ]

  return Trigger({ Context, bindings })
}

export namespace Http {
  export function Get<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'get', options)
  }

  export function Post<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'post', options)
  }

  export function Patch<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'patch', options)
  }

  export function Put<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'put', options)
  }

  export function Delete<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'delete', options)
  }

  export function Head<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'head', options)
  }

  export function Options<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'options', options)
  }

  export function Trace<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'trace', options)
  }

  export function Connect<C extends HttpContext>(route: string, options?: HttpOptions<C>) {
    return Http<C>(route, 'connect', options)
  }

  export function Body(selector?: (body: HttpContext['body'], context: HttpContext) => unknown) {
    if (selector) return Bind((context: HttpContext) => selector(context.body, context))

    return Bind((context: HttpContext) => context.body)
  }

  export function Param(
    nameOrSelector?: string | ((params: HttpContext['params'], context: HttpContext) => unknown),
  ) {
    if (typeof nameOrSelector === 'string')
      return Bind((context: HttpContext) => context.params[nameOrSelector])

    if (typeof nameOrSelector === 'function')
      return Bind((context: HttpContext) => nameOrSelector(context.params, context))

    return Bind((context: HttpContext) => context.params)
  }

  export function Query(
    nameOrSelector?: string | ((query: HttpContext['query'], context: HttpContext) => unknown),
  ) {
    if (typeof nameOrSelector === 'string')
      return Bind((context: HttpContext) => context.query[nameOrSelector])

    if (typeof nameOrSelector === 'function')
      return Bind((context: HttpContext) => nameOrSelector(context.query, context))

    return Bind((context: HttpContext) => context.query)
  }

  export function Status<C extends HttpContext>(status: number) {
    return OnSuccess<C>((payload, context) => {
      context.status(status)
    })
  }

  export function Header<C extends HttpContext>(
    name: string,
    value: string | ((context: C, payload: unknown) => string),
  ): TriggerDecorator
  export function Header(name: string): TriggerParamDecorator
  export function Header<C extends HttpContext>(
    name: string,
    maybeValueOrValueFactory?: string | ((context: C, payload: unknown) => string),
  ) {
    if (maybeValueOrValueFactory === void 0)
      return Bind<C>((context) => context.headers[name.toLowerCase()])

    return OnSuccess<C>((payload, context) => {
      const value =
        typeof maybeValueOrValueFactory === 'function'
          ? maybeValueOrValueFactory(context, payload)
          : maybeValueOrValueFactory

      context.setHeader(name, value)
    }) as TriggerDecorator
  }
}
