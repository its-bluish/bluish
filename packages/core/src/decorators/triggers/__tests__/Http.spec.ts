/* eslint-disable @typescript-eslint/no-magic-numbers */
/* eslint-disable max-lines */
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-unused-vars */
import { HttpContext } from '../../../models/contexts/HttpContext'
import { Runner } from '../../../models/Runner'
import { run } from '../../../test-utils/run'
import { Bind } from '../../Bind'
import { Http } from '../Http'

describe('Http', () => {
  it.each([
    'Get',
    'Post',
    'Patch',
    'Put',
    'Delete',
    'Head',
    'Options',
    'Trace',
    'Connect',
  ] as const)('create a trigger with %s', async (method) => {
    const MethodDecorator: typeof Http.Get = Http[method]
    class Testing {
      @MethodDecorator('/')
      public testing() {
        return method
      }
    }

    await expect(
      run.testing(Testing, void 0, {
        headers: {},
      }),
    ).resolves.toMatchSnapshot()
  })

  describe('.Query', () => {
    it('appends to the first argument taking the entire query object', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Query() query: unknown) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          query: { testing: true },
          headers: {},
        },
      )
      expect(testing).toBeCalledWith({ testing: true })
    })

    it('appends to the first argument by selecting the value by name', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Query('testing') testing: boolean) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          query: { testing: true },
          headers: {},
        },
      )
      expect(testing).toBeCalledWith(true)
    })

    it('appends to the first argument by selecting the value', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Query((query) => query.testing) testing: boolean) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          query: { testing: true },
          headers: {},
        },
      )
      expect(testing).toBeCalledWith(true)
    })
  })

  describe('.Body', () => {
    it('appends to the first argument taking the entire query object', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Body() body: unknown) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          body: { testing: true },
          headers: {},
        },
      )
      expect(testing).toBeCalledWith({ testing: true })
    })

    it('appends to the first argument by selecting the value', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Body((body) => body === 'true') testing: boolean) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          body: 'true',
          headers: {},
        },
      )
      expect(testing).toBeCalledWith(true)
    })
  })

  describe('.Param', () => {
    it('appends to the first argument taking the entire query object', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Param() query: unknown) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          params: { testing: true },
          headers: {},
        },
      )
      expect(testing).toBeCalledWith({ testing: true })
    })

    it('appends to the first argument by selecting the value by name', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Param('testing') testing: boolean) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          params: { testing: true },
          headers: {},
        },
      )
      expect(testing).toBeCalledWith(true)
    })

    it('appends to the first argument by selecting the value', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Param((query) => query.testing) testing: boolean) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          params: { testing: true },
          headers: {},
        },
      )
      expect(testing).toBeCalledWith(true)
    })
  })

  describe('.Header', () => {
    it('appends to the first argument of getting the header value', async () => {
      class Testing {
        @Http.Get('/')
        public testing(@Http.Header('Content-Type') contentType: string) {}
      }
      const testing = jest.spyOn(Testing.prototype, 'testing')
      await run.testing(
        Testing,
        {},
        {
          headers: {
            'Content-Type': 'application/json',
          },
        },
      )
      expect(testing).toBeCalledWith('application/json')
    })

    it('add a header on the trigger result', async () => {
      class Testing {
        @Http.Get('/')
        @Http.Header('Content-Type', 'application/json')
        public testing() {}
      }
      await expect(run.testing(Testing, {}, { headers: {} })).resolves.toMatchSnapshot()
    })
  })

  describe('.Status', () => {
    it('add a status to the requisition', async () => {
      class Testing {
        @Http.Post('/')
        @Http.Status(201)
        public testing() {}
      }
      await expect(run.testing(Testing, {}, { headers: {} })).resolves.toMatchSnapshot()
    })
  })
})
