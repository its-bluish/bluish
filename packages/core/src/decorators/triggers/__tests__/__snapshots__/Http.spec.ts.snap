// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Http .Header add a header on the trigger result 1`] = `
{
  "body": undefined,
  "headers": {
    "Content-Type": "application/json",
  },
  "status": undefined,
}
`;

exports[`Http .Status add a status to the requisition 1`] = `
{
  "body": undefined,
  "headers": {},
  "status": 201,
}
`;

exports[`Http create a trigger with Connect 1`] = `
{
  "body": "Connect",
  "headers": {},
  "status": undefined,
}
`;

exports[`Http create a trigger with Delete 1`] = `
{
  "body": "Delete",
  "headers": {},
  "status": undefined,
}
`;

exports[`Http create a trigger with Get 1`] = `
{
  "body": "Get",
  "headers": {},
  "status": undefined,
}
`;

exports[`Http create a trigger with Head 1`] = `
{
  "body": "Head",
  "headers": {},
  "status": undefined,
}
`;

exports[`Http create a trigger with Options 1`] = `
{
  "body": "Options",
  "headers": {},
  "status": undefined,
}
`;

exports[`Http create a trigger with Patch 1`] = `
{
  "body": "Patch",
  "headers": {},
  "status": undefined,
}
`;

exports[`Http create a trigger with Post 1`] = `
{
  "body": "Post",
  "headers": {},
  "status": undefined,
}
`;

exports[`Http create a trigger with Put 1`] = `
{
  "body": "Put",
  "headers": {},
  "status": undefined,
}
`;

exports[`Http create a trigger with Trace 1`] = `
{
  "body": "Trace",
  "headers": {},
  "status": undefined,
}
`;
